<!DOCTYPE html> <html lang="en"> <head> <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"> <meta charset="utf-8"> <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"> <meta http-equiv="X-UA-Compatible" content="IE=edge"> <title> Long-Read Sequencing Pipeline | Ryan Gallagher </title> <meta name="author" content="Ryan Gallagher"> <meta name="description" content="customized for Whole Genomes on Oxford Nanopore Long-Read Platforms"> <meta name="keywords" content="biostatistics, statistical genetics, genomics, genome assembly, bioinformatics, Ryan Gallagher"> <link rel="stylesheet" href="/assets/css/bootstrap.min.css?a4b3f509e79c54a512b890d73235ef04"> <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/css/mdb.min.css" integrity="sha256-jpjYvU3G3N6nrrBwXJoVEYI/0zw8htfFnhT9ljN3JJw=" crossorigin="anonymous"> <link defer rel="stylesheet" href="/assets/css/academicons.min.css?f0b7046b84e425c55f3463ac249818f5"> <link defer rel="stylesheet" href="/assets/css/scholar-icons.css?62b2ac103a88034e6882a5be5f3e2772"> <link defer rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700|Roboto+Slab:100,300,400,500,700|Material+Icons&amp;display=swap"> <link defer rel="stylesheet" href="/assets/css/jekyll-pygments-themes-github.css?591dab5a4e56573bf4ef7fd332894c99" media="" id="highlight_theme_light"> <link rel="shortcut icon" href="data:image/svg+xml,&lt;svg%20xmlns=%22http://www.w3.org/2000/svg%22%20viewBox=%220%200%20100%20100%22&gt;&lt;text%20y=%22.9em%22%20font-size=%2290%22&gt;%F0%9F%A7%AC&lt;/text&gt;&lt;/svg&gt;"> <link rel="stylesheet" href="/assets/css/main.css?d41d8cd98f00b204e9800998ecf8427e"> <link rel="canonical" href="https://ryan-biostat.github.io/projects/ONT_pipeline/"> <script src="/assets/js/theme.js?a81d82887dd692e91686b43de4542f18"></script> <link defer rel="stylesheet" href="/assets/css/jekyll-pygments-themes-native.css?5847e5ed4a4568527aa6cfab446049ca" media="none" id="highlight_theme_dark"> <script>
    initTheme();
  </script> </head> <body class="fixed-top-nav "> <header> <nav id="navbar" class="navbar navbar-light navbar-expand-sm fixed-top" role="navigation"> <div class="container"> <a class="navbar-brand title font-weight-lighter" href="/"> <span class="font-weight-bold">Ryan</span> Gallagher </a> <button class="navbar-toggler collapsed ml-auto" type="button" data-toggle="collapse" data-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar top-bar"></span> <span class="icon-bar middle-bar"></span> <span class="icon-bar bottom-bar"></span> </button> <div class="collapse navbar-collapse text-right" id="navbarNav"> <ul class="navbar-nav ml-auto flex-nowrap"> <li class="nav-item "> <a class="nav-link" href="/">about </a> </li> <li class="nav-item active"> <a class="nav-link" href="/projects/">projects <span class="sr-only">(current)</span> </a> </li> <li class="nav-item "> <a class="nav-link" href="/cv/">cv </a> </li> <li class="nav-item "> <a class="nav-link" href="/teaching/">teaching </a> </li> <li class="nav-item "> <a class="nav-link" href="/publications/">publications </a> </li> <li class="nav-item "> <a class="nav-link" href="/presentations/">presentations </a> </li> <li class="toggle-container"> <button id="light-toggle" title="Change theme"> <i class="ti ti-sun-moon" id="light-toggle-system"></i> <i class="ti ti-moon-filled" id="light-toggle-dark"></i> <i class="ti ti-sun-filled" id="light-toggle-light"></i> </button> </li> </ul> </div> </div> </nav> <progress id="progress" value="0"> <div class="progress-container"> <span class="progress-bar"></span> </div> </progress> </header> <div class="container mt-5" role="main"> <div class="post"> <header class="post-header"> <h1 class="post-title">Long-Read Sequencing Pipeline</h1> <p class="post-description">customized for Whole Genomes on Oxford Nanopore Long-Read Platforms</p> </header> <article> <h2 id="purpose">Purpose</h2> <p>This pipeline is designed for processing whole genome samples from the Oxford Nanopore Technologies (ONT) PromethION long-read sequencing platform. In our lab, we process these samples on R10.9.1 flowcells, and this workflow interprets the resulting raw sequence data.</p> <p>The pipeline is managed by a central Python script that feeds the data through a series of specialized bioinformatics software to move from raw data to an annotated list of genetic variants.</p> <h2 id="method">Method</h2> <p>The pipeline consists of the following key steps:</p> <h4 id="1-transfer">1. <strong>Transfer</strong> </h4> <p>Files are prepared for cross-server transfer. An <code class="language-plaintext highlighter-rouge">rsync</code> command is used to copy over raw sequencing data from the sequencing instrument to our lab’s High-Performance Computing (HPC) cluster, named “maple”.</p> <h4 id="2-alignment">2. <strong>Alignment</strong> </h4> <p>Raw sequences are aligned to both the hg38 (GRCh38) and T2T (Telomere-to-Telomere) human reference genomes using <code class="language-plaintext highlighter-rouge">minimap2</code>. This process generates a BAM (.bam) file, which serves as the foundational file for subsequent analyses.</p> <h4 id="3-cnv-calling">3. <strong>CNV Calling</strong> </h4> <p><code class="language-plaintext highlighter-rouge">ont-spectre</code>, a Copy Number Variant (CNV) calling tool developed by ONT, is utilized. Although still in development, we use it to create VCF (.vcf) files capable of identifying and annotating large structural variants, specifically those greater than 100kb. This step produces both .vcf and .bed files to aid in analysis.</p> <h4 id="4-sv-calling">4. <strong>SV Calling</strong> </h4> <p><code class="language-plaintext highlighter-rouge">sniffles2</code> is employed as our primary structural variant (SV) caller. It identifies a range of SVs, including deletions, insertions, translocations, and inversions. The typical size range for variants detected by <code class="language-plaintext highlighter-rouge">sniffles2</code> is between 10bp and 50kb. The output is a .vcf file.</p> <h4 id="5-snpsmall-indel-calling">5. <strong>SNP/Small INDEL Calling</strong> </h4> <p>We use <code class="language-plaintext highlighter-rouge">deepvariant</code> from Google’s DeepMind to call small genetic variants. This software utilizes a deep neural network to analyze the sequencing data. This step calls single nucleotide polymorphisms (SNPs) which are the most common type of genetic variation, representing a change in a single DNA building block (nucleotide), and small INsertions and DELetions (INDELs). This analysis also produces haplotyping information, allowing us to determine which genome strand belongs to which parent.</p> <h4 id="6-phasing">6. <strong>Phasing</strong> </h4> <p><code class="language-plaintext highlighter-rouge">whatshap</code> is used to assign haplotypes to our .bam files. This assignment allows us to determine which allele (parental copy) holds which mutations. This is critical for determining the inheritance mode for potential genetic diseases. <code class="language-plaintext highlighter-rouge">whatshap</code> achieves this by taking the information produced by <code class="language-plaintext highlighter-rouge">deepvariant</code> and annotating the .bam file, resulting in a .haplotagged.bam file.</p> <h4 id="7-annotation">7. <strong>Annotation</strong> </h4> <p><code class="language-plaintext highlighter-rouge">snpEFF</code> is used to annotate the small variants (SNPs and INDELs). It enriches the .vcf file by pulling information from multiple public databases, such as ClinVar and dbSNP, to provide aggregate data about known mutations.</p> <p>We also utilize the <strong>Geneyx</strong> software. This is a proprietary annotation platform that adds richer annotations and formats the datasets in a more contestable, browser-accessible way.</p> <h2 id="discussion">Discussion</h2> <p>This pipeline can be executed via the following command:</p> <figure class="highlight"><pre><code class="language-python" data-lang="python"><table class="rouge-table"><tbody><tr>
<td class="gutter gl"><pre class="lineno">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
</pre></td> <td class="code"><pre><span class="c1"># Activate the primary conda environment before running the pipeline
</span><span class="n">conda</span> <span class="n">activate</span> <span class="n">spectre</span>

<span class="n">python</span> <span class="n">pipeline</span><span class="p">.</span><span class="n">py</span> <span class="o">&lt;</span><span class="n">sample_name</span><span class="o">&gt;</span> <span class="o">&lt;</span><span class="n">promethion_data_dir</span><span class="o">&gt;</span> <span class="o">&lt;</span><span class="n">maple_output_dir</span><span class="o">&gt;</span> <span class="p">[</span><span class="o">--</span><span class="n">log</span><span class="o">-</span><span class="n">level</span> <span class="n">LEVEL</span><span class="p">]</span>

<span class="c1">## Help Output
</span>
<span class="n">usage</span><span class="p">:</span> <span class="n">pipeline</span><span class="p">.</span><span class="n">py</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">log</span><span class="o">-</span><span class="n">level</span> <span class="p">{</span><span class="n">DEBUG</span><span class="p">,</span><span class="n">INFO</span><span class="p">,</span><span class="n">WARNING</span><span class="p">,</span><span class="n">ERROR</span><span class="p">,</span><span class="n">CRITICAL</span><span class="p">}]</span>
                   <span class="n">sample_name</span> <span class="n">data_dir</span> <span class="n">to_dir</span>

<span class="n">ONT</span> <span class="n">PromethION</span> <span class="n">Data</span> <span class="n">Processing</span> <span class="n">Pipeline</span><span class="p">.</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="n">sample_name</span>           <span class="n">The</span> <span class="n">name</span> <span class="n">of</span> <span class="n">the</span> <span class="n">sample</span> <span class="n">being</span> <span class="n">processed</span><span class="p">.</span> <span class="n">This</span> <span class="n">should</span>
                        <span class="n">match</span> <span class="n">a</span> <span class="n">created</span> <span class="n">directory</span><span class="p">.</span>
  <span class="n">data_dir</span>              <span class="n">Full</span> <span class="n">path</span> <span class="n">to</span> <span class="n">the</span> <span class="n">directory</span> <span class="n">containing</span> <span class="n">the</span> <span class="n">sample</span> <span class="n">data</span>
                        <span class="n">on</span> <span class="n">the</span> <span class="n">PromethION</span> <span class="nf">server </span><span class="p">(</span><span class="n">e</span><span class="p">.</span><span class="n">g</span><span class="p">.,</span>
                        <span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">run_folder</span><span class="o">/</span><span class="n">sample_subfolder</span><span class="o">/</span><span class="p">).</span>
  <span class="n">to_dir</span>                <span class="n">Full</span> <span class="n">path</span> <span class="n">to</span> <span class="n">the</span> <span class="n">MAPLE</span> <span class="n">directory</span> <span class="n">where</span> <span class="n">output</span> <span class="n">should</span>
                        <span class="n">be</span> <span class="nf">stored </span><span class="p">(</span><span class="n">e</span><span class="p">.</span><span class="n">g</span><span class="p">.,</span> <span class="o">/</span><span class="n">data2</span><span class="o">/</span><span class="n">flowcell_10</span><span class="p">.</span><span class="mf">4.1</span><span class="o">/</span><span class="n">mcw_svi_</span><span class="p">...</span><span class="o">/</span><span class="p">).</span>

<span class="n">options</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="nb">exit</span>
  <span class="o">--</span><span class="n">log</span><span class="o">-</span><span class="n">level</span> <span class="p">{</span><span class="n">DEBUG</span><span class="p">,</span><span class="n">INFO</span><span class="p">,</span><span class="n">WARNING</span><span class="p">,</span><span class="n">ERROR</span><span class="p">,</span><span class="n">CRITICAL</span><span class="p">}</span>
                        <span class="n">Set</span> <span class="n">the</span> <span class="n">logging</span> <span class="nf">level </span><span class="p">(</span><span class="n">default</span><span class="p">:</span> <span class="n">INFO</span><span class="p">)</span>
</pre></td> </tr></tbody></table></code></pre></figure> </article> <h2>References</h2> <div class="publications"> </div> </div> </div> <footer class="fixed-bottom" role="contentinfo"> <div class="container mt-0"> © Copyright 2025 Ryan Gallagher. Powered by <a href="https://jekyllrb.com/" target="_blank" rel="external nofollow noopener">Jekyll</a> with the <a href="https://github.com/alshedivat/al-folio" rel="external nofollow noopener" target="_blank">al-folio</a> theme, hosted on <a href="https://pages.github.com/" target="_blank" rel="external nofollow noopener">GitHub Pages</a>. </div> </footer> <script src="https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.min.js" integrity="sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=" crossorigin="anonymous"></script> <script src="/assets/js/bootstrap.bundle.min.js"></script> <script src="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/js/mdb.min.js" integrity="sha256-NdbiivsvWt7VYCt6hYNT3h/th9vSTL4EDWeGs5SN3DA=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/masonry-layout@4.2.2/dist/masonry.pkgd.min.js" integrity="sha256-Nn1q/fx0H7SNLZMQ5Hw5JLaTRZp0yILA/FRexe19VdI=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/imagesloaded@5.0.0/imagesloaded.pkgd.min.js" integrity="sha256-htrLFfZJ6v5udOG+3kNLINIKh2gvoKqwEhHYfTTMICc=" crossorigin="anonymous"></script> <script defer src="/assets/js/masonry.js?a0db7e5d5c70cc3252b3138b0c91dcaf" type="text/javascript"></script> <script defer src="https://cdn.jsdelivr.net/npm/medium-zoom@1.1.0/dist/medium-zoom.min.js" integrity="sha256-ZgMyDAIYDYGxbcpJcfUnYwNevG/xi9OHKaR/8GK+jWc=" crossorigin="anonymous"></script> <script defer src="/assets/js/zoom.js?85ddb88934d28b74e78031fd54cf8308"></script> <script src="/assets/js/no_defer.js?2781658a0a2b13ed609542042a859126"></script> <script defer src="/assets/js/common.js?c15de51d4bb57887caa2c21988d97279"></script> <script defer src="/assets/js/copy_code.js?c8a01c11a92744d44b093fc3bda915df" type="text/javascript"></script> <script defer src="/assets/js/jupyter_new_tab.js?d9f17b6adc2311cbabd747f4538bb15f"></script> <script async src="https://d1bxh8uas1mnw7.cloudfront.net/assets/embed.js"></script> <script async src="https://badge.dimensions.ai/badge.js"></script> <script defer type="text/javascript" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script> <script src="/assets/js/mathjax-setup.js?a5bb4e6a542c546dd929b24b8b236dfd"></script> <script defer src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6" crossorigin="anonymous"></script> <script defer src="/assets/js/progress-bar.js?2f30e0e6801ea8f5036fa66e1ab0a71a" type="text/javascript"></script> <script src="/assets/js/vanilla-back-to-top.min.js?f40d453793ff4f64e238e420181a1d17"></script> <script>
    addBackToTop();
  </script> </body> </html>