<!DOCTYPE html> <html lang="en"> <head> <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"> <meta charset="utf-8"> <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"> <meta http-equiv="X-UA-Compatible" content="IE=edge"> <title> Learning AlphaGenome | Ryan Gallagher </title> <meta name="author" content="Ryan Gallagher"> <meta name="description" content="investigating Google's newest tool that predicts how genetic mutations in DNA affect gene regulation and other biological processes"> <meta name="keywords" content="biostatistics, statistical genetics, genomics, genome assembly, bioinformatics, Ryan Gallagher"> <link rel="stylesheet" href="/assets/css/bootstrap.min.css?a4b3f509e79c54a512b890d73235ef04"> <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/css/mdb.min.css" integrity="sha256-jpjYvU3G3N6nrrBwXJoVEYI/0zw8htfFnhT9ljN3JJw=" crossorigin="anonymous"> <link defer rel="stylesheet" href="/assets/css/academicons.min.css?f0b7046b84e425c55f3463ac249818f5"> <link defer rel="stylesheet" href="/assets/css/scholar-icons.css?62b2ac103a88034e6882a5be5f3e2772"> <link defer rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700|Roboto+Slab:100,300,400,500,700|Material+Icons&amp;display=swap"> <link defer rel="stylesheet" href="/assets/css/jekyll-pygments-themes-github.css?591dab5a4e56573bf4ef7fd332894c99" media="" id="highlight_theme_light"> <link rel="shortcut icon" href="data:image/svg+xml,&lt;svg%20xmlns=%22http://www.w3.org/2000/svg%22%20viewBox=%220%200%20100%20100%22&gt;&lt;text%20y=%22.9em%22%20font-size=%2290%22&gt;%F0%9F%A7%AC&lt;/text&gt;&lt;/svg&gt;"> <link rel="stylesheet" href="/assets/css/main.css?d41d8cd98f00b204e9800998ecf8427e"> <link rel="canonical" href="https://ryan-biostat.github.io/projects/alphagenome/"> <script src="/assets/js/theme.js?a81d82887dd692e91686b43de4542f18"></script> <link defer rel="stylesheet" href="/assets/css/jekyll-pygments-themes-native.css?5847e5ed4a4568527aa6cfab446049ca" media="none" id="highlight_theme_dark"> <script>
    initTheme();
  </script> </head> <body class="fixed-top-nav "> <header> <nav id="navbar" class="navbar navbar-light navbar-expand-sm fixed-top" role="navigation"> <div class="container"> <a class="navbar-brand title font-weight-lighter" href="/"> <span class="font-weight-bold">Ryan</span> Gallagher </a> <button class="navbar-toggler collapsed ml-auto" type="button" data-toggle="collapse" data-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar top-bar"></span> <span class="icon-bar middle-bar"></span> <span class="icon-bar bottom-bar"></span> </button> <div class="collapse navbar-collapse text-right" id="navbarNav"> <ul class="navbar-nav ml-auto flex-nowrap"> <li class="nav-item "> <a class="nav-link" href="/">about </a> </li> <li class="nav-item active"> <a class="nav-link" href="/projects/">projects <span class="sr-only">(current)</span> </a> </li> <li class="nav-item "> <a class="nav-link" href="https://github.com/ryan-biostat/cv/blob/main/RG_CV_11082025.pdf" rel="external nofollow noopener" target="_blank">cv </a> </li> <li class="nav-item "> <a class="nav-link" href="/teaching/">teaching </a> </li> <li class="nav-item "> <a class="nav-link" href="/presentations/">presentations </a> </li> <li class="nav-item "> <a class="nav-link" href="/publications/">publications </a> </li> <li class="toggle-container"> <button id="light-toggle" title="Change theme"> <i class="ti ti-sun-moon" id="light-toggle-system"></i> <i class="ti ti-moon-filled" id="light-toggle-dark"></i> <i class="ti ti-sun-filled" id="light-toggle-light"></i> </button> </li> </ul> </div> </div> </nav> <progress id="progress" value="0"> <div class="progress-container"> <span class="progress-bar"></span> </div> </progress> </header> <div class="container mt-5" role="main"> <div class="post"> <header class="post-header"> <h1 class="post-title">Learning AlphaGenome</h1> <p class="post-description">investigating Google's newest tool that predicts how genetic mutations in DNA affect gene regulation and other biological processes</p> </header> <article> <h4 id="about"><strong>About</strong></h4> <p>New tools in genomics are very exciting and I have a particular affection for Google’s tools. We currently implement DeepVariant for our small-variant calling, and I haven’t worked with it personally, but I’m aware of how powerful AlphaFold is for protein science. So when I learned that AlphaGenome was a new tool that could apply to our research in genomics, I was very excited to get into the weeds with it. This page documents my time learning and ingesting this very new tool.</p> <h2 id="purpose">Purpose</h2> <p>Our lab is deeply invested in understanding how genetic variation (both large and small) shapes phenotype. AlphaGenome doesn’t directly solve that problem yet, but its ability to approximate the impact of variants on the surrounding genomic context could move us ever-so closer. By building hands-on expertise with AlphaGenome now, we want to be ready for the moment when models of this class can more explicitly link sequence to phenotype. In the meantime, we expect the framework and intermediate outputs to generate useful insights on side questions and adjacent projects, making the time spent learning this tool a clear net positive for the lab.</p> <h2 id="method">Method</h2> <ol> <li> <h4 id="read-the-documentation"><strong>Read the Documentation</strong></h4> <p>The <a href="https://www.alphagenomedocs.com/" rel="external nofollow noopener" target="_blank">AlphaGenome documentation</a> provides a full starting path by walking through obtaining an API key, installing the Python client, and running the model on small, toy examples to verify that predictions can be generated. It also introduces the core abstractions used throughout the API (such as sequence and variant objects) and demonstrates how to submit regions or variants for prediction. The docs emphasize that the current implementation is best suited for &lt;100kb analyses which is a current limitation of the software. It also describes the kind of tissue-data available and the type of variables its capable of predicting.</p> </li> <li> <h4 id="learn-from-community-explanations"><strong>Learn from Community Explanations</strong></h4> <p>As a complement to the official documentation, we also spent time learning from the broader community. The official forums are a great place to get answers to questions and see where people might be similarly struggling. This resource also links out to a lot of great community made videos, which aid heavily for understanding. I found some useful explanation videos about initial install and testing <a href="https://www.youtube.com/watch?v=D_EVxgyd3sc" rel="external nofollow noopener" target="_blank">here</a>. I also learned a lot about the theory behind this tool via a journal club video <a href="https://www.youtube.com/watch?v=ihfQUBgsWxY&amp;t=3296s" rel="external nofollow noopener" target="_blank">here</a>.</p> </li> <li> <h4 id="work-through-tutorials"><strong>Work Through Tutorials</strong></h4> <p>To apply my learning, I setup a Python environment on my MacBook and installed the AlphaGenome client and its dependencies. I followed tutorials for getting API access, and I began working on the “Scoring and Visualizing a Variant” tutorial which walks through extracting a genomic window around a variant and running predictions.</p> </li> </ol> <h2 id="code-sample">Code Sample</h2> <p>My worked tutorial from the documentation:</p> <figure class="highlight"><pre><code class="language-python" data-lang="python"><table class="rouge-table"><tbody><tr>
<td class="gutter gl"><pre class="lineno">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250
251
252
253
254
255
256
257
258
259
260
261
262
263
264
265
266
267
268
269
270
271
272
273
274
275
276
277
278
279
280
281
282
283
284
285
286
287
288
289
290
291
292
293
294
295
296
297
298
299
300
301
302
303
304
305
306
307
308
309
310
311
312
313
314
315
</pre></td> <td class="code"><pre><span class="sh">'''</span><span class="s">
Ryan Gallagher

Scoring and Visualizing a Single Variant

Adapted from https://www.alphagenomedocs.com/colabs/variant_scoring_ui.html

Here, we go through the process of predicting the effect of a single variant on different modalities, such as gene expression and chromatin accessibility.

For my application, I</span><span class="sh">'</span><span class="s">m going to substitute the tutorial variant for a real one that we see in (SAMPLE).

Utilizing Geneyx - we find that there is an intronic deletion on the DEAF1 gene - which seems to match our patient</span><span class="sh">'</span><span class="s">s pheno well.

</span><span class="sh">'''</span>

<span class="kn">from</span> <span class="n">secret.api_key</span> <span class="kn">import</span> <span class="n">ALPHA_GENOME_API_KEY</span>

<span class="kn">from</span> <span class="n">alphagenome.data</span> <span class="kn">import</span> <span class="n">gene_annotation</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">transcript</span>
<span class="kn">from</span> <span class="n">alphagenome.models</span> <span class="kn">import</span> <span class="n">dna_client</span><span class="p">,</span> <span class="n">variant_scorers</span>
<span class="kn">from</span> <span class="n">alphagenome.visualization</span> <span class="kn">import</span> <span class="n">plot_components</span>
<span class="kn">import</span> <span class="n">pandas</span> <span class="k">as</span> <span class="n">pd</span>

<span class="c1"># Load the model
</span><span class="n">dna_model</span> <span class="o">=</span> <span class="n">dna_client</span><span class="p">.</span><span class="nf">create</span><span class="p">(</span><span class="nc">ALPHA_GENOME_API_KEY</span><span class="p">())</span>

<span class="n">HG38_GTF_FEATHER</span> <span class="o">=</span> <span class="p">(</span>
    <span class="sh">'</span><span class="s">https://storage.googleapis.com/alphagenome/reference/gencode/</span><span class="sh">'</span>
    <span class="sh">'</span><span class="s">hg38/gencode.v46.annotation.gtf.gz.feather</span><span class="sh">'</span>
<span class="p">)</span>

<span class="c1"># Initialize an empty directory to serve as a variant effect prediction cache.
</span><span class="n">_prediction_cache</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">_transcript_extractor_cache</span> <span class="o">=</span> <span class="p">{}</span>

<span class="nf">print</span><span class="p">(</span><span class="sh">'</span><span class="se">\n</span><span class="s"> ---- Score Variant ---- </span><span class="se">\n</span><span class="sh">'</span><span class="p">)</span>

<span class="n">organism</span> <span class="o">=</span> <span class="sh">'</span><span class="s">human</span><span class="sh">'</span>
<span class="n">organism_map</span> <span class="o">=</span> <span class="p">{</span><span class="sh">'</span><span class="s">human</span><span class="sh">'</span><span class="p">:</span> <span class="n">dna_client</span><span class="p">.</span><span class="n">Organism</span><span class="p">.</span><span class="n">HOMO_SAPIENS</span><span class="p">}</span>
<span class="n">organism</span> <span class="o">=</span> <span class="n">organism_map</span><span class="p">[</span><span class="n">organism</span><span class="p">]</span>

<span class="c1"># A real deletion on the DEAF1 gene of SVI 0162 UDD
</span><span class="n">variant_chromosome</span> <span class="o">=</span> <span class="sh">'</span><span class="s">chr9</span><span class="sh">'</span>
<span class="n">variant_position</span> <span class="o">=</span> <span class="mi">134644202</span>
<span class="n">variant_reference_bases</span> <span class="o">=</span> <span class="sh">''</span>
<span class="n">variant_alternate_bases</span> <span class="o">=</span> <span class="sh">'</span><span class="s">GGGGAGGGGGCGGCTGTCCACTGGAGATGCAGGCGTGGC</span><span class="sh">'</span>

<span class="n">variant</span> <span class="o">=</span> <span class="n">genome</span><span class="p">.</span><span class="nc">Variant</span><span class="p">(</span>
        <span class="n">chromosome</span> <span class="o">=</span> <span class="n">variant_chromosome</span><span class="p">,</span>
        <span class="n">position</span> <span class="o">=</span> <span class="n">variant_position</span><span class="p">,</span>
        <span class="n">reference_bases</span> <span class="o">=</span> <span class="n">variant_reference_bases</span><span class="p">,</span>
        <span class="n">alternate_bases</span> <span class="o">=</span> <span class="n">variant_alternate_bases</span><span class="p">,</span>
        <span class="p">)</span>

<span class="c1"># Specify length of sequence around variant to predict (this is taking HG38 surrounding the event and predicting when there is vs. isnt a deletion there)
</span>
<span class="n">sequence_length</span> <span class="o">=</span> <span class="sh">'</span><span class="s">1MB</span><span class="sh">'</span> <span class="c1"># options = '2KB', '16KB', '100KB', '500KB', '1MB' as string
</span><span class="n">sequence_length</span> <span class="o">=</span> <span class="n">dna_client</span><span class="p">.</span><span class="n">SUPPORTED_SEQUENCE_LENGTHS</span><span class="p">[</span>
        <span class="sa">f</span><span class="sh">'</span><span class="s">SEQUENCE_LENGTH_</span><span class="si">{</span><span class="n">sequence_length</span><span class="si">}</span><span class="sh">'</span>
        <span class="p">]</span>

<span class="c1"># The input interval is derived from the variant (centered on it).
</span><span class="n">interval</span> <span class="o">=</span> <span class="n">variant</span><span class="p">.</span><span class="n">reference_interval</span><span class="p">.</span><span class="nf">resize</span><span class="p">(</span><span class="n">sequence_length</span><span class="p">)</span>

<span class="c1"># Additonal settings
</span>
<span class="n">variant_scores</span> <span class="o">=</span> <span class="n">dna_model</span><span class="p">.</span><span class="nf">score_variant</span><span class="p">(</span>
        <span class="n">interval</span><span class="o">=</span><span class="n">interval</span><span class="p">,</span>
        <span class="n">variant</span><span class="o">=</span><span class="n">variant</span><span class="p">,</span>
        <span class="n">variant_scorers</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span><span class="n">variant_scorers</span><span class="p">.</span><span class="n">RECOMMENDED_VARIANT_SCORERS</span><span class="p">.</span><span class="nf">values</span><span class="p">()),</span>
        <span class="p">)</span>

<span class="n">df_scores</span> <span class="o">=</span> <span class="n">variant_scorers</span><span class="p">.</span><span class="nf">tidy_scores</span><span class="p">(</span><span class="n">variant_scores</span><span class="p">)</span>

<span class="n">download_predictions</span> <span class="o">=</span> <span class="bp">True</span>
<span class="k">if</span> <span class="n">download_predictions</span><span class="p">:</span>
    <span class="n">df_scores</span><span class="p">.</span><span class="nf">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="sh">'</span><span class="s">chr11_DEAF1_DEL_scores.csv</span><span class="sh">'</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>

<span class="n">columns</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">c</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">df_scores</span><span class="p">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">c</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="sh">'</span><span class="s">variant_id</span><span class="sh">'</span><span class="p">,</span> <span class="sh">'</span><span class="s">scored_intervals</span><span class="sh">'</span><span class="p">]</span>
        <span class="p">]</span>
<span class="c1">#print(df_scores[columns])
</span>

<span class="nf">print</span><span class="p">(</span><span class="sh">'</span><span class="se">\n</span><span class="s"> ---- Visualize Variant Effects ---- </span><span class="se">\n</span><span class="sh">'</span><span class="p">)</span>

<span class="c1"># Specify list of cell and tissue ontologies
</span>
<span class="c1"># Since I don't know a single thing about what I would select here,
# I asked Gemini what cell types I should use given the sample phenotype / gene I'm looking at.
</span>
<span class="c1"># It output Ontology CURIE, and I double checked them from the documentation: https://www.alphagenomedocs.com/colabs/tissue_ontology_mapping.html
</span>
<span class="c1"># OG_ontology_terms = ['UBERON:0002037', 'UBERON:0000955', 'UBERON:0001873'] # Cerebellum, Brain, Caudate Nucleus
</span><span class="n">ontology_terms</span> <span class="o">=</span> <span class="p">[</span><span class="sh">'</span><span class="s">CL:0002551</span><span class="sh">'</span><span class="p">,</span> <span class="sh">'</span><span class="s">CL:0002553</span><span class="sh">'</span><span class="p">,</span> <span class="sh">'</span><span class="s">CL:0002547</span><span class="sh">'</span><span class="p">,</span> <span class="sh">'</span><span class="s">CL:1001608</span><span class="sh">'</span><span class="p">]</span>
<span class="n">plot_gene_annotation</span> <span class="o">=</span> <span class="bp">True</span>
<span class="n">plot_longest_transcript_only</span> <span class="o">=</span> <span class="bp">True</span>

<span class="c1"># Output types to plot:
# There are 13 different output options - each have their own scoring and interpretation
# https://www.alphagenomedocs.com/exploring_model_metadata.html
</span>

<span class="n">plot_rna_seq</span> <span class="o">=</span> <span class="bp">True</span> <span class="c1"># Could be useful if CAGE shows something
</span><span class="n">plot_cage</span> <span class="o">=</span> <span class="bp">True</span>
<span class="n">plot_atac</span> <span class="o">=</span> <span class="bp">False</span>
<span class="n">plot_dnase</span> <span class="o">=</span> <span class="bp">False</span>
<span class="n">plot_chip_histone</span> <span class="o">=</span> <span class="bp">False</span>
<span class="n">plot_chip_tf</span> <span class="o">=</span> <span class="bp">False</span>
<span class="n">plot_splice_sites</span> <span class="o">=</span> <span class="bp">False</span>
<span class="n">plot_splice_site_usage</span> <span class="o">=</span> <span class="bp">False</span>
<span class="n">plot_contact_maps</span> <span class="o">=</span> <span class="bp">False</span>
<span class="n">plot_splice_junctions</span> <span class="o">=</span> <span class="bp">False</span>


<span class="c1"># Options to filter tracks to only a specific DNA strand
## DEAF1 IS ON THE NEGATIVE STRAND
</span><span class="n">filter_to_positive_strand</span> <span class="o">=</span> <span class="bp">False</span>
<span class="n">filter_to_negative_strand</span> <span class="o">=</span> <span class="bp">True</span>

<span class="k">if</span> <span class="n">filter_to_positive_strand</span> <span class="ow">and</span> <span class="n">filter_to_negative_strand</span><span class="p">:</span>
    <span class="k">raise</span> <span class="nc">ValueError</span><span class="p">(</span>
            <span class="sh">'</span><span class="s">Cannot specify both filter_to_positive_strand and </span><span class="sh">'</span>
            <span class="sh">'</span><span class="s">filter_to_negative strand.</span><span class="sh">'</span>
            <span class="p">)</span>

<span class="c1"># Other visualization options:
</span><span class="n">ref_color</span> <span class="o">=</span> <span class="sh">'</span><span class="s">blue</span><span class="sh">'</span>
<span class="n">alt_color</span> <span class="o">=</span> <span class="sh">'</span><span class="s">red</span><span class="sh">'</span>
<span class="n">ref_alt_colors</span> <span class="o">=</span> <span class="p">{</span><span class="sh">'</span><span class="s">REF</span><span class="sh">'</span><span class="p">:</span> <span class="n">ref_color</span><span class="p">,</span> <span class="sh">'</span><span class="s">ALT</span><span class="sh">'</span><span class="p">:</span> <span class="n">alt_color</span><span class="p">}</span>
<span class="n">plot_interval_width</span> <span class="o">=</span> <span class="mi">43008</span>
<span class="n">plot_interval_shift</span> <span class="o">=</span> <span class="mi">0</span>

<span class="c1"># Load gene annotation
</span><span class="k">if</span> <span class="n">organism</span> <span class="ow">in</span> <span class="n">_transcript_extractor_cache</span><span class="p">:</span>
    <span class="n">transcript_extractor</span><span class="p">,</span> <span class="n">longest_transcript_extractor</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">_transcript_extractor_cache</span><span class="p">[</span><span class="n">organism</span><span class="p">]</span>
            <span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="n">match</span> <span class="n">organism</span><span class="p">:</span>
        <span class="n">case</span> <span class="n">dna_client</span><span class="p">.</span><span class="n">Organism</span><span class="p">.</span><span class="n">HOMO_SAPIENS</span><span class="p">:</span>
            <span class="n">gtf_path</span> <span class="o">=</span> <span class="n">HG38_GTF_FEATHER</span>
        <span class="n">case</span> <span class="n">_</span><span class="p">:</span>
            <span class="k">raise</span> <span class="nc">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="sh">'</span><span class="s">Unsupported organism: </span><span class="si">{</span><span class="n">organism</span><span class="si">}</span><span class="sh">'</span><span class="p">)</span>

    <span class="n">gtf</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="nf">read_feather</span><span class="p">(</span><span class="n">gtf_path</span><span class="p">)</span>
    
    <span class="c1"># Filter to protein-coding genes and highly supported transcripts
</span>    <span class="n">gtf_transcript</span> <span class="o">=</span> <span class="n">gene_annotation</span><span class="p">.</span><span class="nf">filter_transcript_support_level</span><span class="p">(</span>
            <span class="n">gene_annotation</span><span class="p">.</span><span class="nf">filter_protein_coding</span><span class="p">(</span><span class="n">gtf</span><span class="p">),</span> <span class="p">[</span><span class="sh">'</span><span class="s">1</span><span class="sh">'</span><span class="p">]</span>
            <span class="p">)</span>
    
    <span class="c1"># Extractor for identifying transcripts in the region
</span>    <span class="n">transcript_extractor</span> <span class="o">=</span> <span class="n">transcript</span><span class="p">.</span><span class="nc">TranscriptExtractor</span><span class="p">(</span><span class="n">gtf_transcript</span><span class="p">)</span>
    
    <span class="c1"># Also define an extractor that fetches only the longest transcript gene.
</span>    <span class="n">gtf_longest_transcript</span> <span class="o">=</span> <span class="n">gene_annotation</span><span class="p">.</span><span class="nf">filter_to_longest_transcript</span><span class="p">(</span>
            <span class="n">gtf_transcript</span>
            <span class="p">)</span>
    
    <span class="n">longest_transcript_extractor</span> <span class="o">=</span> <span class="n">transcript</span><span class="p">.</span><span class="nc">TranscriptExtractor</span><span class="p">(</span>
            <span class="n">gtf_longest_transcript</span>
            <span class="p">)</span>
    <span class="n">_transcript_extractor_cache</span><span class="p">[</span><span class="n">organism</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">transcript_extractor</span><span class="p">,</span>
            <span class="n">longest_transcript_extractor</span>
            <span class="p">)</span>

<span class="k">def</span> <span class="nf">_predict_variant_cached</span><span class="p">(</span>
    <span class="n">interval</span><span class="p">,</span> <span class="n">variant</span><span class="p">,</span> <span class="n">organism</span><span class="p">,</span> <span class="n">requested_outputs</span><span class="p">,</span> <span class="n">ontology_terms</span>
<span class="p">):</span>
  <span class="sh">"""</span><span class="s">Cache wrapper of dna_model.predict_variant.</span><span class="sh">"""</span>
  <span class="c1"># Create a unique key from the function arguments.
</span>  <span class="n">cache_key</span> <span class="o">=</span> <span class="p">(</span>
      <span class="nf">str</span><span class="p">(</span><span class="n">interval</span><span class="p">),</span>
      <span class="nf">str</span><span class="p">(</span><span class="n">variant</span><span class="p">),</span>
      <span class="nf">str</span><span class="p">(</span><span class="n">organism</span><span class="p">),</span>
      <span class="nf">tuple</span><span class="p">(</span><span class="n">requested_outputs</span><span class="p">),</span>
      <span class="nf">tuple</span><span class="p">(</span><span class="n">ontology_terms</span><span class="p">),</span>
  <span class="p">)</span>

  <span class="c1"># Check if the result is already in the cache.
</span>  <span class="k">if</span> <span class="n">cache_key</span> <span class="ow">in</span> <span class="n">_prediction_cache</span><span class="p">:</span>
    <span class="k">return</span> <span class="n">_prediction_cache</span><span class="p">[</span><span class="n">cache_key</span><span class="p">]</span>

  <span class="c1"># If not, compute the prediction and store it in the cache.
</span>  <span class="n">result</span> <span class="o">=</span> <span class="n">dna_model</span><span class="p">.</span><span class="nf">predict_variant</span><span class="p">(</span>
      <span class="n">interval</span><span class="o">=</span><span class="n">interval</span><span class="p">,</span>
      <span class="n">variant</span><span class="o">=</span><span class="n">variant</span><span class="p">,</span>
      <span class="n">organism</span><span class="o">=</span><span class="n">organism</span><span class="p">,</span>
      <span class="n">requested_outputs</span><span class="o">=</span><span class="n">requested_outputs</span><span class="p">,</span>
      <span class="n">ontology_terms</span><span class="o">=</span><span class="n">ontology_terms</span><span class="p">,</span>
  <span class="p">)</span>
  <span class="n">_prediction_cache</span><span class="p">[</span><span class="n">cache_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">result</span>
  <span class="k">return</span> <span class="n">result</span>

<span class="n">output</span> <span class="o">=</span> <span class="nf">_predict_variant_cached</span><span class="p">(</span>
        <span class="n">interval</span> <span class="o">=</span> <span class="n">interval</span><span class="p">,</span>
        <span class="n">variant</span> <span class="o">=</span> <span class="n">variant</span><span class="p">,</span>
        <span class="n">organism</span> <span class="o">=</span> <span class="n">organism</span><span class="p">,</span>
        <span class="n">requested_outputs</span> <span class="o">=</span> <span class="p">[</span><span class="o">*</span><span class="n">dna_client</span><span class="p">.</span><span class="n">OutputType</span><span class="p">],</span>
        <span class="n">ontology_terms</span> <span class="o">=</span> <span class="n">ontology_terms</span><span class="p">,</span>
        <span class="p">)</span>

<span class="c1"># Filter to DNA strand if requested
</span><span class="n">ref</span><span class="p">,</span> <span class="n">alt</span> <span class="o">=</span> <span class="n">output</span><span class="p">.</span><span class="n">reference</span><span class="p">,</span> <span class="n">output</span><span class="p">.</span><span class="n">alternate</span>

<span class="k">if</span> <span class="n">filter_to_positive_strand</span><span class="p">:</span>
    <span class="n">ref</span> <span class="o">=</span> <span class="n">ref</span><span class="p">.</span><span class="nf">filter_to_strand</span><span class="p">(</span><span class="n">strand</span> <span class="o">=</span> <span class="sh">"</span><span class="s">+</span><span class="sh">"</span><span class="p">)</span>
    <span class="n">alt</span> <span class="o">=</span> <span class="n">alt</span><span class="p">.</span><span class="nf">filter_to_strand</span><span class="p">(</span><span class="n">strand</span> <span class="o">=</span> <span class="sh">"</span><span class="s">+</span><span class="sh">"</span><span class="p">)</span>
<span class="k">elif</span> <span class="n">filter_to_negative_strand</span><span class="p">:</span>
    <span class="n">ref</span> <span class="o">=</span> <span class="n">ref</span><span class="p">.</span><span class="nf">filter_to_strand</span><span class="p">(</span><span class="n">strand</span><span class="o">=</span><span class="sh">'</span><span class="s">-</span><span class="sh">'</span><span class="p">)</span>
    <span class="n">alt</span> <span class="o">=</span> <span class="n">alt</span><span class="p">.</span><span class="nf">filter_to_strand</span><span class="p">(</span><span class="n">strand</span><span class="o">=</span><span class="sh">'</span><span class="s">-</span><span class="sh">'</span><span class="p">)</span>


<span class="c1"># Build plot.
</span><span class="n">components</span> <span class="o">=</span> <span class="p">[]</span>

<span class="c1"># Gene and transcript annotation
</span><span class="k">if</span> <span class="n">plot_gene_annotation</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">plot_longest_transcript_only</span><span class="p">:</span>
        <span class="n">transcripts</span> <span class="o">=</span> <span class="n">longest_transcript_extractor</span><span class="p">.</span><span class="nf">extract</span><span class="p">(</span><span class="n">interval</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">transcripts</span> <span class="o">=</span> <span class="n">transcript_extractor</span><span class="p">.</span><span class="nf">extract</span><span class="p">(</span><span class="n">interval</span><span class="p">)</span>
    <span class="n">components</span><span class="p">.</span><span class="nf">append</span><span class="p">(</span><span class="n">plot_components</span><span class="p">.</span><span class="nc">TranscriptAnnotation</span><span class="p">(</span><span class="n">transcripts</span><span class="p">))</span>


<span class="c1"># Individual output type plots.
</span><span class="n">plot_map</span> <span class="o">=</span> <span class="p">{</span>
        <span class="sh">'</span><span class="s">plot_atac</span><span class="sh">'</span><span class="p">:</span> <span class="p">(</span><span class="n">ref</span><span class="p">.</span><span class="n">atac</span><span class="p">,</span> <span class="n">alt</span><span class="p">.</span><span class="n">atac</span><span class="p">,</span> <span class="sh">'</span><span class="s">ATAC</span><span class="sh">'</span><span class="p">),</span>
        <span class="sh">'</span><span class="s">plot_cage</span><span class="sh">'</span><span class="p">:</span> <span class="p">(</span><span class="n">ref</span><span class="p">.</span><span class="n">cage</span><span class="p">,</span> <span class="n">alt</span><span class="p">.</span><span class="n">cage</span><span class="p">,</span> <span class="sh">'</span><span class="s">CAGE</span><span class="sh">'</span><span class="p">),</span>
        <span class="sh">'</span><span class="s">plot_chip_histone</span><span class="sh">'</span><span class="p">:</span> <span class="p">(</span><span class="n">ref</span><span class="p">.</span><span class="n">chip_histone</span><span class="p">,</span> <span class="n">alt</span><span class="p">.</span><span class="n">chip_histone</span><span class="p">,</span> <span class="sh">'</span><span class="s">CHIP_HISTONE</span><span class="sh">'</span><span class="p">),</span>
        <span class="sh">'</span><span class="s">plot_chip_tf</span><span class="sh">'</span><span class="p">:</span> <span class="p">(</span><span class="n">ref</span><span class="p">.</span><span class="n">chip_tf</span><span class="p">,</span> <span class="n">alt</span><span class="p">.</span><span class="n">chip_tf</span><span class="p">,</span> <span class="sh">'</span><span class="s">CHIP_TF</span><span class="sh">'</span><span class="p">),</span>
        <span class="sh">'</span><span class="s">plot_contact_maps</span><span class="sh">'</span><span class="p">:</span> <span class="p">(</span><span class="n">ref</span><span class="p">.</span><span class="n">contact_maps</span><span class="p">,</span> <span class="n">alt</span><span class="p">.</span><span class="n">contact_maps</span><span class="p">,</span> <span class="sh">'</span><span class="s">CONTACT_MAPS</span><span class="sh">'</span><span class="p">),</span>
        <span class="sh">'</span><span class="s">plot_dnase</span><span class="sh">'</span><span class="p">:</span> <span class="p">(</span><span class="n">ref</span><span class="p">.</span><span class="n">dnase</span><span class="p">,</span> <span class="n">alt</span><span class="p">.</span><span class="n">dnase</span><span class="p">,</span> <span class="sh">'</span><span class="s">DNASE</span><span class="sh">'</span><span class="p">),</span>
        <span class="sh">'</span><span class="s">plot_rna_seq</span><span class="sh">'</span><span class="p">:</span> <span class="p">(</span><span class="n">ref</span><span class="p">.</span><span class="n">rna_seq</span><span class="p">,</span> <span class="n">alt</span><span class="p">.</span><span class="n">rna_seq</span><span class="p">,</span> <span class="sh">'</span><span class="s">RNA_SEQ</span><span class="sh">'</span><span class="p">),</span>
        <span class="sh">'</span><span class="s">plot_splice_junctions</span><span class="sh">'</span><span class="p">:</span> <span class="p">(</span>
            <span class="n">ref</span><span class="p">.</span><span class="n">splice_junctions</span><span class="p">,</span>
            <span class="n">alt</span><span class="p">.</span><span class="n">splice_junctions</span><span class="p">,</span>
            <span class="sh">'</span><span class="s">SPLICE_JUNCTIONS</span><span class="sh">'</span><span class="p">,</span>
        <span class="p">),</span>
        <span class="sh">'</span><span class="s">plot_splice_sites</span><span class="sh">'</span><span class="p">:</span> <span class="p">(</span><span class="n">ref</span><span class="p">.</span><span class="n">splice_sites</span><span class="p">,</span> <span class="n">alt</span><span class="p">.</span><span class="n">splice_sites</span><span class="p">,</span> <span class="sh">'</span><span class="s">SPLICE_SITES</span><span class="sh">'</span><span class="p">),</span>
        <span class="sh">'</span><span class="s">plot_splice_site_usage</span><span class="sh">'</span><span class="p">:</span> <span class="p">(</span>
            <span class="n">ref</span><span class="p">.</span><span class="n">splice_site_usage</span><span class="p">,</span>
            <span class="n">alt</span><span class="p">.</span><span class="n">splice_site_usage</span><span class="p">,</span>
            <span class="sh">'</span><span class="s">SPLICE_SITE_USAGE</span><span class="sh">'</span><span class="p">,</span>
    <span class="p">),</span>
<span class="p">}</span>

<span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="p">(</span><span class="n">ref_data</span><span class="p">,</span> <span class="n">alt_data</span><span class="p">,</span> <span class="n">output_type</span><span class="p">)</span> <span class="ow">in</span> <span class="n">plot_map</span><span class="p">.</span><span class="nf">items</span><span class="p">():</span>
  <span class="k">if</span> <span class="nf">eval</span><span class="p">(</span><span class="n">key</span><span class="p">)</span> <span class="ow">and</span> <span class="n">ref_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span> <span class="ow">and</span> <span class="n">ref_data</span><span class="p">.</span><span class="n">values</span><span class="p">.</span><span class="n">shape</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
    <span class="nf">print</span><span class="p">(</span>
        <span class="sa">f</span><span class="sh">'</span><span class="s">Requested plot for output </span><span class="si">{</span><span class="n">output_type</span><span class="si">}</span><span class="s"> but no tracks exist in</span><span class="sh">'</span>
        <span class="sh">'</span><span class="s"> output. This is likely because this output does not exist for your</span><span class="sh">'</span>
        <span class="sh">'</span><span class="s"> ontologies or requested DNA strand.</span><span class="sh">'</span>
    <span class="p">)</span>
  <span class="k">if</span> <span class="nf">eval</span><span class="p">(</span><span class="n">key</span><span class="p">)</span> <span class="ow">and</span> <span class="n">ref_data</span> <span class="ow">and</span> <span class="n">alt_data</span><span class="p">:</span>
    <span class="n">match</span> <span class="n">output_type</span><span class="p">:</span>
      <span class="n">case</span> <span class="sh">'</span><span class="s">CHIP_HISTONE</span><span class="sh">'</span><span class="p">:</span>
        <span class="n">ylabel_template</span> <span class="o">=</span> <span class="p">(</span>
            <span class="sa">f</span><span class="sh">'</span><span class="si">{</span><span class="n">output_type</span><span class="si">}</span><span class="s">:  ()</span><span class="se">\n</span><span class="sh">'</span>
        <span class="p">)</span>
      <span class="n">case</span> <span class="sh">'</span><span class="s">CHIP_TF</span><span class="sh">'</span><span class="p">:</span>
        <span class="n">ylabel_template</span> <span class="o">=</span> <span class="p">(</span>
            <span class="sa">f</span><span class="sh">'</span><span class="si">{</span><span class="n">output_type</span><span class="si">}</span><span class="s">: </span><span class="sh">'</span>
            <span class="sh">'</span><span class="s"> ({strand})</span><span class="se">\n</span><span class="s">{transcription_factor}</span><span class="sh">'</span>
        <span class="p">)</span>
      <span class="n">case</span> <span class="sh">'</span><span class="s">CONTACT_MAPS</span><span class="sh">'</span><span class="p">:</span>
        <span class="n">ylabel_template</span> <span class="o">=</span> <span class="sa">f</span><span class="sh">'</span><span class="si">{</span><span class="n">output_type</span><span class="si">}</span><span class="s">:  ()</span><span class="sh">'</span>
      <span class="n">case</span> <span class="sh">'</span><span class="s">SPLICE_SITES</span><span class="sh">'</span><span class="p">:</span>
        <span class="n">ylabel_template</span> <span class="o">=</span> <span class="sa">f</span><span class="sh">'</span><span class="si">{</span><span class="n">output_type</span><span class="si">}</span><span class="s">:  ()</span><span class="sh">'</span>
      <span class="n">case</span> <span class="n">_</span><span class="p">:</span>
        <span class="n">ylabel_template</span> <span class="o">=</span> <span class="p">(</span>
            <span class="sa">f</span><span class="sh">'</span><span class="si">{</span><span class="n">output_type</span><span class="si">}</span><span class="s">:  ()</span><span class="se">\n</span><span class="sh">'</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">output_type</span> <span class="o">==</span> <span class="sh">'</span><span class="s">CONTACT_MAPS</span><span class="sh">'</span><span class="p">:</span>
      <span class="n">component</span> <span class="o">=</span> <span class="n">plot_components</span><span class="p">.</span><span class="nc">ContactMapsDiff</span><span class="p">(</span>
          <span class="n">tdata</span><span class="o">=</span><span class="n">alt_data</span> <span class="o">-</span> <span class="n">ref_data</span><span class="p">,</span>
          <span class="n">ylabel_template</span><span class="o">=</span><span class="n">ylabel_template</span><span class="p">,</span>
      <span class="p">)</span>
      <span class="n">components</span><span class="p">.</span><span class="nf">append</span><span class="p">(</span><span class="n">component</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">output_type</span> <span class="o">==</span> <span class="sh">'</span><span class="s">SPLICE_JUNCTIONS</span><span class="sh">'</span><span class="p">:</span>
      <span class="n">ref_plot</span> <span class="o">=</span> <span class="n">plot_components</span><span class="p">.</span><span class="nc">Sashimi</span><span class="p">(</span>
          <span class="n">ref_data</span><span class="p">,</span>
          <span class="n">ylabel_template</span><span class="o">=</span><span class="sh">'</span><span class="s">REF: </span><span class="sh">'</span> <span class="o">+</span> <span class="n">ylabel_template</span><span class="p">,</span>
      <span class="p">)</span>
      <span class="n">alt_plot</span> <span class="o">=</span> <span class="n">plot_components</span><span class="p">.</span><span class="nc">Sashimi</span><span class="p">(</span>
          <span class="n">alt_data</span><span class="p">,</span>
          <span class="n">ylabel_template</span><span class="o">=</span><span class="sh">'</span><span class="s">ALT: </span><span class="sh">'</span> <span class="o">+</span> <span class="n">ylabel_template</span><span class="p">,</span>
      <span class="p">)</span>
      <span class="n">components</span><span class="p">.</span><span class="nf">extend</span><span class="p">([</span><span class="n">ref_plot</span><span class="p">,</span> <span class="n">alt_plot</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="n">component</span> <span class="o">=</span> <span class="n">plot_components</span><span class="p">.</span><span class="nc">OverlaidTracks</span><span class="p">(</span>
          <span class="n">tdata</span><span class="o">=</span><span class="p">{</span><span class="sh">'</span><span class="s">REF</span><span class="sh">'</span><span class="p">:</span> <span class="n">ref_data</span><span class="p">,</span> <span class="sh">'</span><span class="s">ALT</span><span class="sh">'</span><span class="p">:</span> <span class="n">alt_data</span><span class="p">},</span>
          <span class="n">colors</span><span class="o">=</span><span class="n">ref_alt_colors</span><span class="p">,</span>
          <span class="n">ylabel_template</span><span class="o">=</span><span class="n">ylabel_template</span><span class="p">,</span>
          <span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.6</span> <span class="c1">#ADDED FOR TRANSPARENCY
</span>      <span class="p">)</span>
      <span class="n">components</span><span class="p">.</span><span class="nf">append</span><span class="p">(</span><span class="n">component</span><span class="p">)</span>

<span class="k">if</span> <span class="n">plot_interval_width</span> <span class="o">&gt;</span> <span class="n">interval</span><span class="p">.</span><span class="n">width</span><span class="p">:</span>
  <span class="k">raise</span> <span class="nc">ValueError</span><span class="p">(</span>
      <span class="sa">f</span><span class="sh">'</span><span class="s">plot_interval_width (</span><span class="si">{</span><span class="n">plot_interval_width</span><span class="si">}</span><span class="s">) must be less than </span><span class="sh">'</span>
      <span class="sa">f</span><span class="sh">'</span><span class="s">interval.width (</span><span class="si">{</span><span class="n">interval</span><span class="p">.</span><span class="n">width</span><span class="si">}</span><span class="s">).</span><span class="sh">'</span>
  <span class="p">)</span>

<span class="n">plot</span> <span class="o">=</span> <span class="n">plot_components</span><span class="p">.</span><span class="nf">plot</span><span class="p">(</span>
    <span class="n">components</span><span class="o">=</span><span class="n">components</span><span class="p">,</span>
    <span class="n">interval</span><span class="o">=</span><span class="n">interval</span><span class="p">.</span><span class="nf">shift</span><span class="p">(</span><span class="n">plot_interval_shift</span><span class="p">).</span><span class="nf">resize</span><span class="p">(</span><span class="n">plot_interval_width</span><span class="p">),</span>
    <span class="n">annotations</span><span class="o">=</span><span class="p">[</span>
        <span class="n">plot_components</span><span class="p">.</span><span class="nc">VariantAnnotation</span><span class="p">([</span><span class="n">variant</span><span class="p">]),</span>
    <span class="p">],</span>
<span class="p">)</span>

<span class="n">plot</span><span class="p">.</span><span class="nf">savefig</span><span class="p">(</span><span class="sh">'</span><span class="s">COL5A1_variant_plot.png</span><span class="sh">'</span><span class="p">)</span>
</pre></td> </tr></tbody></table></code></pre></figure> </article> </div> </div> <footer class="fixed-bottom" role="contentinfo"> <div class="container mt-0"> © Copyright 2025 Ryan Gallagher. Powered by <a href="https://jekyllrb.com/" target="_blank" rel="external nofollow noopener">Jekyll</a> with the <a href="https://github.com/alshedivat/al-folio" rel="external nofollow noopener" target="_blank">al-folio</a> theme, hosted on <a href="https://pages.github.com/" target="_blank" rel="external nofollow noopener">GitHub Pages</a>. </div> </footer> <script src="https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.min.js" integrity="sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=" crossorigin="anonymous"></script> <script src="/assets/js/bootstrap.bundle.min.js"></script> <script src="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/js/mdb.min.js" integrity="sha256-NdbiivsvWt7VYCt6hYNT3h/th9vSTL4EDWeGs5SN3DA=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/masonry-layout@4.2.2/dist/masonry.pkgd.min.js" integrity="sha256-Nn1q/fx0H7SNLZMQ5Hw5JLaTRZp0yILA/FRexe19VdI=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/imagesloaded@5.0.0/imagesloaded.pkgd.min.js" integrity="sha256-htrLFfZJ6v5udOG+3kNLINIKh2gvoKqwEhHYfTTMICc=" crossorigin="anonymous"></script> <script defer src="/assets/js/masonry.js?a0db7e5d5c70cc3252b3138b0c91dcaf" type="text/javascript"></script> <script defer src="https://cdn.jsdelivr.net/npm/medium-zoom@1.1.0/dist/medium-zoom.min.js" integrity="sha256-ZgMyDAIYDYGxbcpJcfUnYwNevG/xi9OHKaR/8GK+jWc=" crossorigin="anonymous"></script> <script defer src="/assets/js/zoom.js?85ddb88934d28b74e78031fd54cf8308"></script> <script src="/assets/js/no_defer.js?2781658a0a2b13ed609542042a859126"></script> <script defer src="/assets/js/common.js?c15de51d4bb57887caa2c21988d97279"></script> <script defer src="/assets/js/copy_code.js?c8a01c11a92744d44b093fc3bda915df" type="text/javascript"></script> <script defer src="/assets/js/jupyter_new_tab.js?d9f17b6adc2311cbabd747f4538bb15f"></script> <script async src="https://d1bxh8uas1mnw7.cloudfront.net/assets/embed.js"></script> <script async src="https://badge.dimensions.ai/badge.js"></script> <script defer type="text/javascript" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script> <script src="/assets/js/mathjax-setup.js?a5bb4e6a542c546dd929b24b8b236dfd"></script> <script defer src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6" crossorigin="anonymous"></script> <script defer src="/assets/js/progress-bar.js?2f30e0e6801ea8f5036fa66e1ab0a71a" type="text/javascript"></script> <script src="/assets/js/vanilla-back-to-top.min.js?f40d453793ff4f64e238e420181a1d17"></script> <script>
    addBackToTop();
  </script> </body> </html>